### PR TITLE
New: `template-tag-spacing` rule (fixes #7631)

### LINKED ISSUE #7631: Rule proposal: Template tag spacing
**Please describe what the rule should do:** Similar to func-call-spacing – it would enforce spacing between template tag and template literal.

**What category of rule is this? (place an "X" next to just one item)**

- [X] Enforces code style
- [ ] Warns about a potential error
- [ ] Suggests an alternate way of doing something
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// correct with spacing required
const [user] = await sql `SELECT * FROM users WHERE id = ${id}`
ctx.body = {
  name: user.name,
  avatar: escape `/avatars/${user.id}.jpg`,
}

// correct with spacing disallowed
const [user] = await sql`SELECT * FROM users WHERE id = ${id}`
ctx.body = {
  name: user.name,
  avatar: escape`/avatars/${user.id}.jpg`,
}
```

### PR DESCRIPTION
**Pull request by jwilsson**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ X ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**Please describe what the rule should do:**
This rule is very similiar to `func-call-spacing` in the way it will enforce spacing between [template tags](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Tagged_template_literals) and their corresponding template literal.

**What category of rule is this? (place an "X" next to just one item)**

[ X ] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
/* eslint template-tag-spacing: ["error", "never"] */
func `foo`
```

```js
/* eslint template-tag-spacing: ["error", "always"] */
func`foo`
```

**Why should this rule be included in ESLint (instead of a plugin)?**
Since it's really similar to `func-call-spacing` which is already included in ESLint it makes sense to enforce this styling too, especially since tagged template literals will become more and more common.

### CODE REVIEW

File: lib/rules/template-tag-spacing.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "require or disallow spacing between template tags and their literals",
+            category: "ECMAScript 6",
```
Comment (not-an-aardvark) [inline lib/rules/template-tag-spacing.js]: I'm not sure, but I think we're generally trying to move stuff out of the "ECMAScript 6" category (also see https://github.com/eslint/eslint.github.io/issues/310). In my opinion, it might be better to put this in "Stylistic Issues".

File: lib/rules/template-tag-spacing.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+        /**
+         * Check if a space is present between a template tag and its literal
+         * @param {ASTNode} node node to evaluate
+         * @returns {void}
+         * @private
+         */
+        function checkSpacing(node) {
+            const tagToken = sourceCode.getLastToken(node.tag);
+            const literalToken = sourceCode.getTokenAfter(tagToken);
```
Comment (platinumazure) [inline lib/rules/template-tag-spacing.js]: @not-an-aardvark Won't the first line in your example also spit out the `)` token? Should that be `sourceCode.getFirstToken(node.tag)`?

File: conf/eslint.json [hunk 3]
```diff
@@ -231,6 +231,7 @@
         "strict": "off",
         "symbol-description": "off",
         "template-curly-spacing": "off",
+        "template-tag-spacing": "off",
```
Comment (platinumazure) [inline conf/eslint.json]: @not-an-aardvark Should that be in a separate (chore) PR?

Comment (not-an-aardvark) [inline conf/eslint.json]: I would be fine with either.

File: lib/rules/template-tag-spacing.js (hunk 2 above)
Comment (not-an-aardvark) [inline lib/rules/template-tag-spacing.js]: I'm not sure I understood you correctly, but I think we do want the `)` token in this case:

```js
// we care about this space ↓
(          foo          )        `bar`
 // not this one ↑
```

File: lib/rules/template-tag-spacing.js (hunk 1 above)
Comment (vitorbal) [inline lib/rules/template-tag-spacing.js]: "Stylistic issues" sounds good for me as well.

File: conf/eslint.json (hunk 3 above)
Comment (vitorbal) [inline conf/eslint.json]: Good idea on the dog fooding! But agreed with @platinumazure that it should probably be a separate PR

File: docs/rules/template-tag-spacing.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
+# Enforce Usage of Spacing in Tagged Template Literals (template-tag-spacing)
+
+(fixable) The `--fix` option on the [command line](../user-guide/command-line-interface#fix) automatically fixes problems reported by this rule.
+
+We can create functions whose parameters consist of template literals.
```
Comment (jwilsson) [inline docs/rules/template-tag-spacing.md]: Yeah, you're right. I rewrote the whole first paragraph and added a "Further reading" section to explain the whole concept a bit better.

File: conf/eslint.json (hunk 3 above)
Comment (jwilsson) [inline conf/eslint.json]: I'd be happy to submit a chore PR once this one is finished.

**Comment by jwilsson** [comment]
I made the suggested changes. I rewrote the first paragraph of the documentation as well, to better explain what "tagged template literals" is.

File: docs/rules/template-tag-spacing.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
+# Enforce Usage of Spacing in Tagged Template Literals (template-tag-spacing)
+
+(fixable) The `--fix` option on the [command line](../user-guide/command-line-interface#fix) automatically fixes problems reported by this rule.
+
+With ES6, it's possible to create functions called [tagged template literals](#further-reading) where the function parameters consist of a template literal's strings and expressions.
+
+This rule can force usage of spacing _between_ the tag function and the template literal.
```
Comment (nzakas) [inline docs/rules/template-tag-spacing.md]: This section shouldn't be about the rule, it should just be about the concept the rule relates to. So, can you remove this line and do a bit more explanation around what a tagged template is where spacing is allowed? I'd like to see a complete example with the function definition.

**Comment by jwilsson** [comment]
@nzakas I updated the docs, I hope I understood you correctly.

File: docs/rules/template-tag-spacing.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
+# Enforce Usage of Spacing in Tagged Template Literals (template-tag-spacing)
```
Comment (kaicataldo) [inline docs/rules/template-tag-spacing.md]: Can you update this to match [this description](https://github.com/eslint/eslint/pull/7913/files#diff-8dbaf8148ca54cc6af73a2dd5accc3b0R15)?

i.e. `Require or disallow spacing between template tags and their literals (template-tag-spacing)`

File: docs/rules/template-tag-spacing.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+## Options
+
+```json
+{
+    "template-tag-spacing": ["error", "never"]
+}
+```
+
+This rule has one option which has either `"never"` or `"always"` as value.
```
Comment (kaicataldo) [inline docs/rules/template-tag-spacing.md]: Looks like there's a word missing here - maybe `This rule has one option whose value can be set to "never" or "always"`?

**Comment by jwilsson** [comment]
@kaicataldo Done and done!

File: tests/lib/rules/template-tag-spacing.js [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        {
+            code: "tag`name`",
+            output: "tag `name`",
+            errors: [
+                { message: "Missing space between template tag and template literal." }
+            ],
+            options: ["always"]
+        },
+        {
+            code: "tag /*here's a comment*/`Hello world`",
```
Comment (kaicataldo) [inline tests/lib/rules/template-tag-spacing.js]: Thanks for updating this! One final thing - this still gets autofixed so we do need an `output` property (even if it's just to assert that it doesn't get fixed).  With the current logic, I believe it will remove the comment, correct? My preference would be to either figure out a way to keep the comment in the fixed output or to not fix if a comment exists (with a strong preference for the former), rather than simply remove the comment. Thoughts? 

@not-an-aardvark Can you be a resource here? I think you've written more autofixes than the rest of us combined by now!

Comment (not-an-aardvark) [inline tests/lib/rules/template-tag-spacing.js]: I think the best way to fix this would be to get a list of all the comments between the tag and the template literal, then concatenate them to get the resulting text between the two tokens. This would preserve the whitespace in the comments themselves and remove any other whitespace.

However, we should make sure we don't change the behavior of this code:

```js
foo // line comment
`bar`
```

The implementation I mentioned above would fix that to

```js
foo// line comment`bar`
```

To avoid this, I think we should avoid doing a fix if there are any line comments between the tag and the template literal.

So I would do something along the lines of

```js
context.report({
  // ...
  fix(fixer) {
    const comments = sourceCode.getComments(node.quasi).leading;

    if (comments.some(comment => comment.type === "Line")) {
        return null;
    }
    return fixer.replaceTextRange(
        [tagToken.range[1], literalToken.range[0]],
        comments.reduce((text, comment) => text + sourceCode.getText(comment), "")
    );
  }
});
```

Comment (jwilsson) [inline tests/lib/rules/template-tag-spacing.js]: Ah, of course. Sorry about that!

But I'm guessing we shouldn't try to fix anything in cases like this:
```js
/* eslint template-tag-spacing: ["error", "always"] */
tag/*here's a comment*/`Hello world`
```

Since we wouldn't know where to put the space.

@not-an-aardvark Thanks for the explanation and example!

Comment (not-an-aardvark) [inline tests/lib/rules/template-tag-spacing.js]: @jwilsson That seems reasonable. Another option would be to just always put the space immediately after the tag.

Comment (jwilsson) [inline tests/lib/rules/template-tag-spacing.js]: Alright, I added the expected output to all tests and opted for the "space immediately after tag" approach when autofixing comments.

**Comment by not-an-aardvark** [comment]
Thanks for contributing!
